subprojects/execution/src/main/java/org/gradle/internal/execution/fingerprint/impl/DefaultInputFingerprinter.java

```diff
@@ -112,10 +113,14 @@ public void visitInputProperty(String propertyName, ValueSupplier value) {
                     valueSnapshotsBuilder.put(propertyName, valueSnapshotter.snapshot(actualValue, previousSnapshot));
                 }
             } catch (Exception e) {
+                String msg = "value '%s' cannot be serialized";
+                if (actualValue instanceof RegularFile) {
+                    msg += ".  This property might have to use @InputFile, or a related file-based input annotation, instead of @Input";
+                }
```

Comment:
There should probably be some check elsewhere that detects that someone is using `@Input` on a `RegularFileProperty` instead of guessing here.  I think you'd have the same problem with `DirectoryProperty`?

Reply:
Just added a test for using `@Input` instead of `@InputDirectory` on a `DirectoryProperty` that confirms it shows the same behavior.  That case does not print this suggestion in the error message though, it needs a separate check added, as the `actualValue` when using `DirectoryProperty` is not a `RegularFile` (and the message should suggest using `@InputDirectory` in that case anyways).

Following the call path down from `TaskExecution#visitRegularInputs` it is only when using `@Input` that we reach this point, so I don't think it's guessing on the annotation type.  I agree that this sort of check seems like it ought be done earlier in the input-visiting process, I'm just not sure where.